topology/probes/nsm/connection.go

```diff
@@ -0,0 +1,297 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package nsm
+
+import (
+	"fmt"
+
+	localconn "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/local/connection"
+	remoteconn "github.com/networkservicemesh/networkservicemesh/controlplane/pkg/apis/remote/connection"
+	"github.com/skydive-project/skydive/filters"
+	"github.com/skydive-project/skydive/graffiti/graph"
+	"github.com/skydive-project/skydive/logging"
+)
+
+type connection interface {
+	AddEdge(*graph.Graph)
+	DelEdge(*graph.Graph)
+	GetSource() *localconn.Connection
+	GetDest() *localconn.Connection
+	GetInodes() (int64, int64)
+	createMetadatas() graph.Metadata
+}
+
+type baseConnectionPair struct {
+	payload  string
+	srcInode int64
+	dstInode int64
+	src      *localconn.Connection
+	dst      *localconn.Connection
+}
+
+func (b *baseConnectionPair) GetSource() *localconn.Connection {
+	return b.src
+}
+
+func (b *baseConnectionPair) GetDest() *localconn.Connection {
+	return b.dst
+}
+
+func (b *baseConnectionPair) GetSourceInode() int64 {
+	if b.src == nil {
+		return 0
+	}
+	i, err := getLocalInode(b.src)
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func (b *baseConnectionPair) GetDestInode() int64 {
+	if b.dst == nil {
+		return 0
+	}
+	i, err := getLocalInode(b.dst)
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func (b *baseConnectionPair) GetInodes() (int64, int64) {
+	return b.GetSourceInode(), b.GetDestInode()
+}
+
+// A local connection is composed of only one cross-connect
+type localConnectionPair struct {
+	baseConnectionPair
+	ID string // crossConnectID
+}
+
+// A remote connection is composed of two cross-connects
+type remoteConnectionPair struct {
+	baseConnectionPair
+	remote *remoteconn.Connection // the remote connection shared between the two corss-connects
+	srcID  string                 // The id of the cross-connect with a local connection as source
+	dstID  string                 // The id of the cross-connect with a local connection as destination
+
+}
+
+// easyjson:json
+type baseConnectionMetadata struct {
+	MechanismType       string
+	MechanismParameters map[string]string
+	Labels              map[string]string
+}
+
+// easyjson:json
+type localConnectionMetadata struct {
+	IP string
+	baseConnectionMetadata
+}
+
+// easyjson:json
+type remoteConnectionMetadata struct {
+	baseConnectionMetadata
```

Comment:
nit: please put baseConnectionMetadata at the beginning

Reply:
it is already at the beginnig, am I missing something?

Reply:
My bad for this one..